cmd/otelcol/main.go

```diff
@@ -193,16 +235,54 @@ func checkMemorySettingsMiBFromEnvVar(envVar string) int {
 			log.Fatalf("Expected a number greater than 0 for %s env variable but got %s", envVar, envVarVal)
 		}
 		envVarResult = val
-	} else {
+	case memTotalSizeMiB > 0:
+		break
+	default:
 		log.Printf("Usage: %s=12345 %s=us0 %s=684 %s=1024 %s=256 %s", tokenEnvVarName, realmEnvVarName, ballastEnvVarName, memLimitMiBEnvVarName, memSpikeMiBEnvVarName, os.Args[0])
 		log.Fatalf("ERROR: Missing environment variable %s", envVar)
 	}
 	return envVarResult
 }
 
-func useMemorySettingsMiBFromEnvVar() {
-	memLimit := checkMemorySettingsMiBFromEnvVar(memLimitMiBEnvVarName)
-	memSpike := checkMemorySettingsMiBFromEnvVar(memSpikeMiBEnvVarName)
+func useMemorySettingsMiBFromEnvVar(memTotalSizeMiB int) {
+	// Check if memory limit is specified via environment variable
+	memLimit := checkMemorySettingsMiBFromEnvVar(memLimitMiBEnvVarName, memTotalSizeMiB)
+	// Use if set, otherwise memory total size must be specified
+	if memLimit == 0 {
+		if memTotalSizeMiB == 0 {
+			panic("PANIC: Both memory limit MiB and memory total size are set to zero. This should never happen.")
+		}
+		// If not set, compute based on memory total size specified
+		// and default memory limit percentage const
+		memLimitMiB := memTotalSizeMiB * defaultMemoryLimitPercentage / 100
+		// The memory limit should be set to defaultMemoryLimitPercentage of total memory
+		// while reserving a maximum of defaultMemoryLimitMaxMiB of memory.
```

Comment:
I am not sure I understand why we do this. So we specify let's say 8GB of total memory size, but this calculation will give only 2GB to memory limiter? Are we assuming the remaining 6GB will be used by something else?
Just to clarify: the memory limiter calculates the total size of the memory used by the process. I am confused about what we are trying to achieve here.

Reply:
OK, re-reading the code it appears this code tries to limit the unused memory. The unused memory is the 10% which we did not give to memory limiter despite the user telling us to do so.
Is this 10% reserved because we do not trust the memory to enforce it? And so we believe that the memory limiter may overshoot sometimes but never by more than 2GB, that's why if 10% is more than 2GB, we only reserve 2GB but no more.
Is this what we do?